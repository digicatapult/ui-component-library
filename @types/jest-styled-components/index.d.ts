declare module 'jest-styled-components/serializer'
