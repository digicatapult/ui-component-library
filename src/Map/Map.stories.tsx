import React from 'react'

import Map from './index'
import exampleJson from './example.json'
import colors from '../colors'

export default {
  title: 'Map Component',
  component: Map,
  argTypes: {
    sourceJson: {
      control: 'text',
    },
    style: {
      control: 'text',
    },
    height: {
      control: 'text',
    },
    width: {
      control: 'text',
    },
    startPositionLong: {
      control: 'number',
      table: {
        category: 'Start Position',
      },
    },
    startPositionLat: {
      control: 'number',
      table: {
        category: 'Start Position',
      },
    },
    startPositionZoom: {
      control: 'number',
      table: {
        category: 'Start Position',
      },
    },
    cluster: {
      control: 'boolean',
      table: {
        category: 'Clustering',
      },
    },
    clusterMaxZoom: {
      control: 'number',
      table: {
        category: 'Clustering',
      },
    },
    clusterAreaRadius: {
      control: 'number',
      table: {
        category: 'Clustering',
      },
    },
    clusterColor: {
      control: 'color',
      table: {
        category: 'Clustering',
      },
    },
    clusterRadius: {
      control: 'number',
      table: {
        category: 'Clustering',
      },
    },
    countFont: {
      control: 'array',
      table: {
        category: 'Clustering',
      },
    },
    countFontSize: {
      control: 'number',
      table: {
        category: 'Clustering',
      },
    },
    countFontColor: {
      control: 'color',
      table: {
        category: 'Clustering',
      },
    },
    pointColor: {
      control: 'color',
      table: {
        category: 'Clustering',
      },
    },
    pointRadius: {
      control: 'number',
      table: {
        category: 'Clustering',
      },
    },
  },
}

const Template = (args) => {
  const props = {
    token: process.env.STORYBOOK_MAPBOX_TOKEN,
    size: {
      height: args.height,
      width: args.width,
    },
    startPosition: {
      long: args.startPositionLong,
      lat: args.startPositionLat,
      zoom: args.startPositionZoom,
    },
    style: args.style,
    sourceJson: args.sourceJson,
    cluster: args.cluster,
    clusterOptions: {
      clusterMaxZoom: args.clusterMaxZoom,
      clusterAreaRadius: args.clusterAreaRadius,
      clusterColor: args.clusterColor,
      clusterRadius: args.clusterRadius,
      countFont: args.countFont,
      countFontSize: args.countFontSize,
      countFontColor: args.countFontColor,
      pointColor: args.pointColor,
      pointRadius: args.pointRadius,
    },
  }

  return <Map {...props} />
}

export const Cluster = Template.bind({})
Cluster.args = {
  sourceJson: exampleJson,
  cluster: true,
  clusterColor: colors.green,
  clusterRadius: 14,
  pointColor: colors.green,
  pointRadius: 4,
}

export const NoCluster = Template.bind({})
NoCluster.args = {
  sourceJson: exampleJson,
}
