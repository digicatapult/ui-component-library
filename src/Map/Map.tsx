import React, { useRef, useEffect, useState } from 'react'
import styled from 'styled-components'
import mapboxgl from 'mapbox-gl'

import colors from '../colors'

interface StartPosition {
  long: number
  lat: number
  zoom: number
}

interface Size {
  height: string
  width: string
}

interface ClusterOptions {
  clusterMaxZoom: number
  clusterAreaRadius: number
  clusterColor: string
  clusterRadius: number
  countFont: string[]
  countFontSize: number
  countFontColor: string
  pointColor: string
  pointRadius: number
}

export interface Props {
  token: string
  size?: Size
  startPosition?: StartPosition
  style?: string
  sourceJson?: string
  cluster: boolean
  clusterOptions?: ClusterOptions
}

const Wrapper = styled('div')<Size>`
  height: ${({ height }) => height};
  width: ${({ width }) => width};
`

const Map: React.FC<Props> = (props) => {
  const mapContainer = useRef(null)
  const map = useRef<mapboxgl.Map | null>(null)

  mapboxgl.accessToken = props.token

  const height = props.size?.height || '800px'
  const width = props.size?.width || '800px'

  const long = props.startPosition?.long || -3.5
  const lat = props.startPosition?.lat || 55
  const zoom = props.startPosition?.zoom || 5

  const style = props.style || 'mapbox://styles/mapbox/light-v11'
  const source = props.sourceJson || ''
  const cluster = props.cluster || false

  const clusterMaxZoom = props.clusterOptions?.clusterMaxZoom || 14
  const clusterAreaRadius = props.clusterOptions?.clusterAreaRadius || 50
  const clusterColor = props.clusterOptions?.clusterColor || colors.black
  const clusterRadius = props.clusterOptions?.clusterRadius || 20
  const countFont = props.clusterOptions?.countFont || ['Open Sans Regular']
  const countFontSize = props.clusterOptions?.countFontSize || 14
  const countFontColor = props.clusterOptions?.countFontColor || colors.white
  const pointColor = props.clusterOptions?.pointColor || colors.black
  const pointRadius = props.clusterOptions?.pointRadius || 6

  useEffect(() => {
    if (map.current) return undefined // initialize map only once
    map.current = new mapboxgl.Map({
      container: mapContainer.current!,
      style: style,
      center: [long, lat],
      zoom: zoom,
      attributionControl: false,
    })
  })

  useEffect(() => {
    if (!map.current) return undefined // wait for map to initialize
    map.current.on('load', () => {
      map.current?.addSource('source', {
        type: 'geojson',
        data: source,
        cluster: cluster,
        clusterMaxZoom: clusterMaxZoom, // Max zoom to cluster points on
        clusterRadius: clusterAreaRadius, // Radius of each cluster when clustering points (defaults to 50)
      })

      map.current?.addLayer({
        id: 'clusters',
        type: 'circle',
        source: 'source',
        filter: ['has', 'point_count'],
        paint: {
          'circle-color': clusterColor,
          'circle-radius': clusterRadius,
        },
      })

      map.current?.addLayer({
        id: 'cluster-count',
        type: 'symbol',
        source: 'source',
        filter: ['has', 'point_count'],
        layout: {
          'text-field': ['get', 'point_count_abbreviated'],
          'text-font': countFont,
          'text-size': countFontSize,
        },
        paint: {
          'text-color': countFontColor,
        },
      })

      map.current?.addLayer({
        id: 'unclustered-point',
        type: 'circle',
        source: 'source',
        filter: ['!', ['has', 'point_count']],
        paint: {
          'circle-color': pointColor,
          'circle-radius': pointRadius,
        },
      })

      // inspect a cluster on click
      map.current?.on('click', 'clusters', (e) => {
        const features = map.current!.queryRenderedFeatures(e.point, {
          layers: ['clusters'],
        })

        const clusterId = features[0]?.properties?.cluster_id
        const source: mapboxgl.GeoJSONSource = map.current!.getSource(
          'source'
        ) as mapboxgl.GeoJSONSource
        source.getClusterExpansionZoom(clusterId, (err, zoom) => {
          if (err) return

          if (features[0].geometry.type === 'Point') {
            const coordinates = features[0].geometry.coordinates
            map.current?.easeTo({
              center: coordinates as mapboxgl.LngLatLike,
              zoom: zoom,
            })
          }
        })
      })

      map.current?.on('mouseenter', 'clusters', () => {
        map.current!.getCanvas().style.cursor = 'pointer'
      })
      map.current?.on('mouseleave', 'clusters', () => {
        map.current!.getCanvas().style.cursor = ''
      })
    })
  })

  return <Wrapper ref={mapContainer} height={height} width={width}></Wrapper>
}

export default Map
