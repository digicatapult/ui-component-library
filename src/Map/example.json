{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [-1.23644, 54.57304]
      },
      "properties": {}
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [-0.07132, 53.578133]
      },
      "properties": {}
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [-0.2502445, 53.6333774]
      },
      "properties": {}
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [-0.255333, 53.653999]
      },
      "properties": {}
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [-1.23644, 54.57304]
      },
      "properties": {}
    }
  ]
}
