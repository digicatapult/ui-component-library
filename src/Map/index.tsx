import React, { useRef, useEffect } from 'react'
import styled from 'styled-components'
import mapboxgl, {
  LngLatLike,
  MapboxGeoJSONFeature,
  GeoJSONSource,
  Expression,
} from 'mapbox-gl'
import { FeatureCollection, GeoJSON } from 'geojson'

import colors from '../colors.js'

export interface InitialState {
  long?: number
  lat?: number
  zoom?: number
  height?: string
  width?: string
  style?: string
}

export interface ClusterOptions {
  clusterMaxZoom?: number
  clusterAreaRadius?: number
  clusterColor?: string
  clusterRadius?: number
  countFont?: string[]
  countFontSize?: number
  countFontColor?: string
}

export interface PointOptions {
  pointExpression?: Expression
  pointColor?: string
  pointRadius?: number
  onPointClick?: (feature: MapboxGeoJSONFeature) => void
  onClickZoomIn?: number
}

export interface Props {
  token: string
  sourceJson?: GeoJSON
  initialState?: InitialState
  cluster?: boolean
  clusterOptions?: ClusterOptions
  pointOptions?: PointOptions
}

const Wrapper = styled('div')<InitialState>`
  height: ${({ height }) => height};
  width: ${({ width }) => width};
`

const applyLayerDefaults = (props: Props) => {
  return {
    cluster: props.cluster || false,
    clusterOptions: {
      clusterMaxZoom: props.clusterOptions?.clusterMaxZoom || 14,
      clusterAreaRadius: props.clusterOptions?.clusterAreaRadius || 50,
      clusterColor: props.clusterOptions?.clusterColor || colors.black,
      clusterRadius: props.clusterOptions?.clusterRadius || 20,
      countFont: props.clusterOptions?.countFont || ['Open Sans Regular'],
      countFontSize: props.clusterOptions?.countFontSize || 14,
      countFontColor: props.clusterOptions?.countFontColor || colors.white,
    },
    pointOptions: {
      pointExpression: props.pointOptions?.pointExpression || null,
      pointColor: props.pointOptions?.pointColor || colors.black,
      pointRadius: props.pointOptions?.pointRadius || 6,
      onPointClick: props.pointOptions?.onPointClick || Function(),
      onClickZoomIn: props.pointOptions?.onClickZoomIn || 12,
    },
  }
}

const Map: React.FC<Props> = (props) => {
  const mapContainer = useRef(null)
  const mapRef = useRef<mapboxgl.Map | null>(null)

  const {
    cluster,
    clusterOptions: {
      clusterMaxZoom,
      clusterAreaRadius,
      clusterColor,
      clusterRadius,
      countFont,
      countFontSize,
      countFontColor,
    },
    pointOptions: {
      pointExpression,
      pointColor,
      pointRadius,
      onPointClick,
      onClickZoomIn,
    },
  } = applyLayerDefaults(props)

  const sourceJson = props.sourceJson
  const height = props.initialState?.height || '800px'
  const width = props.initialState?.width || '800px'

  mapboxgl.accessToken = props.token

  // initialize map
  useEffect(() => {
    if (mapRef.current) return undefined

    // defaults
    const long = props.initialState?.long || -3.5
    const lat = props.initialState?.lat || 55
    const zoom = props.initialState?.zoom || 5
    const style =
      props.initialState?.style || 'mapbox://styles/mapbox/light-v11'

    mapRef.current = new mapboxgl.Map({
      container: mapContainer.current!,
      style: style,
      center: [long, lat],
      zoom: zoom,
      attributionControl: false,
      testMode: true,
    })

    // clean up
    return () => {
      mapRef?.current?.remove()
      mapRef.current = null
    }
  }, [
    props.initialState?.style,
    props.initialState?.long,
    props.initialState?.lat,
    props.initialState?.zoom,
    // height/width changes require map reload
    props.initialState?.height,
    props.initialState?.width,
  ])

  // update map geojson without map reload
  useEffect(() => {
    const map = mapRef.current
    if (!map || !sourceJson) return undefined

    const geojsonSource = map.getSource('source') as GeoJSONSource
    geojsonSource?.setData(sourceJson as FeatureCollection)
  }, [sourceJson])

  // add layers after map load
  useEffect(() => {
    const map = mapRef.current
    if (!map) return undefined

    map.on('load', () => {
      map.addSource('source', {
        type: 'geojson',
        data: sourceJson,
        cluster: cluster,
        clusterMaxZoom: clusterMaxZoom, // Max zoom to cluster points on
        clusterRadius: clusterAreaRadius, // Radius of each cluster when clustering points
      })

      map.addLayer({
        id: 'clusters',
        type: 'circle',
        source: 'source',
        filter: ['has', 'point_count'],
        paint: {
          'circle-color': clusterColor,
          'circle-radius': clusterRadius,
        },
      })

      map.addLayer({
        id: 'cluster-count',
        type: 'symbol',
        source: 'source',
        filter: ['has', 'point_count'],
        layout: {
          'text-field': ['get', 'point_count_abbreviated'],
          'text-font': countFont,
          'text-size': countFontSize,
        },
        paint: {
          'text-color': countFontColor,
        },
      })

      map.addLayer({
        id: 'unclustered-point',
        type: 'circle',
        source: 'source',
        filter: ['!', ['has', 'point_count']],
        paint: {
          'circle-color':
            pointExpression != null ? pointExpression : pointColor,
          'circle-radius': pointRadius,
        },
      })

      // inspect a cluster on click
      map.on('click', 'clusters', (e) => {
        const features = map.queryRenderedFeatures(e.point, {
          layers: ['clusters'],
        })

        const clusterId = features[0]?.properties?.cluster_id
        const source: mapboxgl.GeoJSONSource = map.getSource(
          'source'
        ) as mapboxgl.GeoJSONSource
        source.getClusterExpansionZoom(clusterId, (err, zoom) => {
          if (err) return

          if (features[0].geometry.type === 'Point') {
            map.easeTo({
              center: features[0].geometry.coordinates as LngLatLike,
              zoom: zoom,
            })
          }
        })
      })

      map.on('click', 'unclustered-point', (e) => {
        const features = map.queryRenderedFeatures(e.point)
        if (features[0].geometry.type === 'Point') {
          onPointClick(features[0])

          // center on point, zoom in if current zoom is less than onClickZoomIn
          map.easeTo({
            center: features[0].geometry.coordinates as LngLatLike,
            zoom: Math.max(map.getZoom(), onClickZoomIn),
          })
        }
      })

      map.on('mouseenter', 'clusters', () => {
        map.getCanvas().style.cursor = 'pointer'
      })
      map.on('mouseleave', 'clusters', () => {
        map.getCanvas().style.cursor = ''
      })
    })
  })

  return <Wrapper ref={mapContainer} height={height} width={width}></Wrapper>
}

export default Map
