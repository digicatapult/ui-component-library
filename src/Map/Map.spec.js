import React from 'react'
import renderer, { act } from 'react-test-renderer'
import mapboxgl from 'mapbox-gl'

import 'jest-styled-components'

import Map from './index.tsx'

jest.mock('mapbox-gl')

const mockMapOn = jest.fn()

mapboxgl.Map.mockImplementation(() => {
  return { on: mockMapOn }
})

describe('Map', () => {
  test('simple', async () => {
    let instance
    await act(() => {
      instance = renderer.create(<Map />)
    })

    expect(mockMapOn).toHaveBeenCalledTimes(1)
    expect(instance.toJSON()).toMatchSnapshot()
  })
})
