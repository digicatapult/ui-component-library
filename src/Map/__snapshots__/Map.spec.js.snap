// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map simple 1`] = `
.c0 {
  height: 800px;
  width: 800px;
}

<div
  className="c0"
  height="800px"
  width="800px"
/>
`;
